In [1]:
file = open("doc_partie2/exemple.ppm")
print(file.read())
file.close()

P3  
# Informations utiles sur l'image 
# (lieu, date de creation,...)
3 2          
255        
255   0   0 
0 255   0  
0   0 255  
0   0   0 
255 255 255  
122  53 201  



In [6]:
def get_picture(filename:str)->list[list[list[int]]] :
    for k in range(5) :
        filename.readline()
    res = [[filename.readline().split() for i in range(3)] for j in range(2)]
    return res
file = open("doc_partie2/exemple.ppm","r")
print(get_picture(file))
file.close()

[[['255', '0', '0'], ['0', '255', '0'], ['0', '0', '255']], [['0', '0', '0'], ['255', '255', '255'], ['122', '53', '201']]]


$\star\underline{\textbf{Fonction write_picture()}}$

In [3]:
def write_picture(img : list[list[list[int]]], comment : str, filename : str) -> None :
    # On ouvre le flux en "écriture" grâce à la fonction open()
    file = open(filename,"w")
    # Ensuite on écrit les informations que doivent contenir l'image notamment la taille,le nombre de pixels
    file.write("P3\n")
    file.write(comment)
    file.write("\n" + str(len(img[0])) + " " + str(len(img)) + "\n")
    file.write("\n"+str(255)+"\n")
    # Grâce à cette double boucle, on pourra parcourir entièrement le tableau 2D et pouvoir attribuer les couleurs à attribuer
    for i in range(len(img)) :
           for j in range(len(img[0])) :
                file.write(str(img[i][j][0])+ " " + str(img[i][j][1]) + " " + str(img[i][j][2])+ "\n" )
                

    #On fait juste un print() pour spécifier l'enregistrement de l'image sous le nom de de la variable 'filename'
    print("Image enregistrée sous",filename,"avec succès")
    # On ferme le flux
    file.close()


$\textbf{I)}\underline{\textbf{ENSEMBLE DE MANDELBROT}}\\ $ 
D'abord une fonction :

 $a)\underline{\textbf{Fonction mandelbrot_sub()}}$ 

### Au préalable on importe le module $cmath$
Cette fonction a essentiellement pour but de définir le principe de voire la base de la construction de l'ensemble de Mandelbrot
Elle prend en paramètre un nombre complexe c et le nombre d'itérations max_it. Ensuite elle fait une évaluation de la suite $(z_n)$ définie par $z_{n+1}=z_n²+c$ et qui a pour premier terme $z_0 = 0$. Alors,elle compare le module de ce complexe avec la valeur 2 si le module est supérieur ou égale à 2 elle renvoie l'indice sinon elle renvoie le nombre maximale d'itérations atteints.

$b)\underline{\textbf{Fonction mandelbrot_set()}}$ 

 Cette fonction prend en paramètre les entiers height et width, qui sont respectivement le nombre de pixels en hauteur et en largeur ensuite les réels $x_{start}\,y_{start}\,x_{end}\,y_{end}\$ qui constituent respectivement les coordonnées des  points de départs et de fin donc à chaque pixel pour la constitution de l'image.


Au premier abord on définit un tableau vide qui constitue l'image en une seule couleur. Dans notre cas l'image est en tout noir.
Ensuite dans une double boucle on détermine les coordonnées de chaque pixel qui constitueront la valeur du nombre complexe $c = real+i* imag$ par la formule suivante :
$real = x * \frac{(x_{end}-x_{start})}{(width-1)}+x_{start}$ et
$imag = y * \frac{(y_{end}-y_{start})}{(height-1)}+y_{start}$
 Ensuite on vérifie on obtient le résultat de construction à travers la vérification et le résultat faite par la fonction $\textbf{mandelbrot_sub()}$. On fait par la suite une comparaison en le résultat obtenu et le nombre maximale d'itération afin d'obtenire le tableau de couleur qu'on désire.
 Ensuite on grâce à la fonction get_color(),on retourne le tableau avec "l'image de k"


In [14]:
import cmath
#Fonction mandelbrot_sub()    
def mandelbrot_sub(c : complex, max_it : int) -> int :
    z = 0
    for i in range(max_it):
        z = z**2 + c
        if abs(z) >= 2 :
            return i
    return max_it

# Fonction mandelbrot_set()
def mandelbrot_set(height:int,width:int,x_start:float,x_end:float,y_start:float,y_end:float) :
#Création de l'image avec la couleur noir
    pict_mandelbrot = [[[0,0,0] for i in range(width)] for j in range(height)]
    
    for y in range(height):
        for x in range(width) :
            real = (x * (x_end-x_start)/(width-1))+x_start
            imag = (y * (y_end-y_start)/(height-1))+y_start
            c = complex(real,imag)
            n = mandelbrot_sub(c,maxi_it)
            #Affichage de l'ensemble 
            if n < maxi_it :
                pict_mandelbrot[y][x] = [255,255,255]
    return pict_mandelbrot
#Entrée des valeurs     
h = 800
w = 800
x_1 = -2
y_1 = -2
x_2 = 2
y_2 = 2
maxi_it = 100

#création de l'image et de l'affichage en blanc noir grâce à la fonction write_picture()
tab = mandelbrot_set(h,w,x_1,x_2,y_1,y_2)
write_picture(tab,"#Mandel","Mandelbrot.ppm")

Image enregistrée sous Mandelbrot.ppm avec succès


 $\textbf{II)}\underline{\textbf{ENSEMBLE DE JULIA}}\\ $ 
L'ensemble de julia est pratiquement similaire à celui de Mandelbrot mais à juste une différence près la variable.

$a)\underline{\textbf{Fonction julia_sub()}}$ 

Cette fonction en différence de la fonction $\textbf{mandelbrot_sub()}$ prend en plus de du complexe c et le nombre d'itérations en paramètre prend aussi le complexe z en paramètre. 
Le corps de cette fonction se représente comme la suite récurrente suivante :
$(z_n)_{n\in\mathbb{N}}$ définie par $z_{n}=z_n²+c$.
Dans le cas de cet ensemble au lieu du complexe c qui varie c'est la variable z qui varie c'est-à-dire au lieu de $c = real+i* imag$ on a $z = real+i* imag$


$b)\underline{\textbf{Fonction julia_set()}}$ 

Elle est exactement similaire à celle de ${\textbf{mandelbrot_set()}}$

$c)\underline{\textbf{AFFICHAGE DE L'ENSEMBLE}}$ 

On obtient l'image correspondante grâce à la fonction $\textbf{write_picture()}$

In [5]:
import cmath
#Fonction julia_sub()
def julia_sub(c : complex,z:complex, max_it : int) -> int :
    for i in range(max_it):
        z = z**2 + c
        if abs(z) >= 2 :
            return i
    return max_it
#Fonction julia_set()
def julia_set(height:int,width:int,x_start:float,x_end:float,y_start:float,y_end:float) :
    #Image en tout noir
    pict_julia = [[[0,0,0] for i in range(width)] for j in range(height)]
    
    for y in range(height):
        for x in range(width) :
            real = x * (x_end-x_start)/(width-1)+x_start
            imag = y * (y_end-y_start)/(height-1)+y_start
            z = complex(real,imag)
            n = julia_sub(c,z,maxi_it)
            if n < maxi_it :
                pict_julia[y][x] = [255,255,255]
    return pict_julia

height = 1000
width = 1000
x_1 = -2
y_1 = -2
x_2 = 2
y_2 = 2
maxi_it = 300
c = complex(-0.8,0.156)

#Affichage en blanc noir de l'ensemble de l'ensemble de Julia 
tab = julia_set(height,width,x_1,x_2,y_1,y_2)
write_picture(tab,"#Julia","Julia.ppm")

Image enregistrée sous Julia.ppm avec succès


$\textbf{III)}$$\underline{\textbf{Fonction get_color() et interpolation linéaire pour l'affichage en couleur}}$


La fonction $\textbf{get_color()}$ a pour principal objectif de pouvoir créer une échelle de couleurs avec interpollation afin de pouvoir mieux oserver les détails de notre fractale. Pour ce faire nous suivre l'exemple qui a été donné dans le document.

In [18]:
def get_color(q : float) -> list[int] :
    if q == 0.0 :
        return [0, 7, 100]
    elif 0<=q and q<0.16 :
        t = q/0.16
        color1 = [0, 7, 100]
        color2 = [32,107,203]
    elif q<=0.16 and q<0.42 :
        t = (q-0.16)/(0.42-0.16)
        color1 = [32,107,203]
        color2 = [237,255,255]
    elif q<=0.42 and q<0.6425 :
        t = (q-0.42)/(0.6425-0.42)
        color1 = [237,255,255]
        color2 = [255, 170, 0]
    elif q<=0.6425 and q<0.8575 :
        t = (q-0.6425)/(0.8575-0.6425)
        color1 = [255, 170, 0]
        color2 = [153, 89, 0]
    else:
        t = (q-0.8575)/(1-0.8575)
        color1 = [153,89,0]
        color2 = [0,2,0]
    res = [int((1-t)*color1[i] + t*color2[i]) for i in range(3)]
    return res


Alors pour utiliser notre fonction $\textbf{get_color()}$,reprenons notre fonction $\textbf{mandelbrot_set()}$ et $\textbf{julia_set()}$ en définissant une variable $k = \frac{n}{maxi_{it}}$ dont on déterminera l'image par $\textbf{get_color()}$ et ensuite au lieu de retourner un tableau blanc on retourne le tableau otenu par cette dernière.

$\star\underline{\textbf{Mandelbrot}}$

In [67]:
def mandelbrot_set(height:int,width:int,x_start:float,x_end:float,y_start:float,y_end:float) :
#Création de l'image avec la couleur noir
    pict_mandelbrot = [[[0,0,0] for i in range(width)] for j in range(height)]
    
    for y in range(height):
        for x in range(width) :
            real = (x * (x_end-x_start)/(width-1))+x_start
            imag = (y * (y_end-y_start)/(height-1))+y_start
            c = complex(real,imag)
            n = mandelbrot_sub(c,maxi_it)
            #Définition de la variable k
            k = n/maxi_it
            #Affichage de l'ensemble 
            if n < maxi_it :
                pict_mandelbrot[y][x] = get_color(k)
    return pict_mandelbrot
#Entrée des valeurs     
h = 1000
w = 1000
x_1 = -2
y_1 = -2
x_2 = 2
y_2 = 2
maxi_it = 100

#création de l'image et de l'affichage en blanc noir grâce à la fonction write_picture()
tab = mandelbrot_set(h,w,x_1,x_2,y_1,y_2)
write_picture(tab,"#Mandel","Mandelbrot_en_couleur.ppm")

Image enregistrée sous Mandelbrot_en_couleur.ppm avec succès
Le programme a mis 10.60235047340393 secondes pour s'exécuter


$\star\underline{\textbf{Julia}}$

In [16]:
def julia_set(height:int,width:int,x_start:float,x_end:float,y_start:float,y_end:float) :
    #Image en tout noir
    pict_julia = [[[0,0,0] for i in range(width)] for j in range(height)]
    
    for y in range(height):
        for x in range(width) :
            real = x * (x_end-x_start)/(width-1)+x_start
            imag = y * (y_end-y_start)/(height-1)+y_start
            z = complex(real,imag)
            n = julia_sub(c,z,maxi_it)
            #Ici au lieu de k, j'ai préféré mettre m pour une légère distinction
            m = n/maxi_it
            if n < maxi_it :
                pict_julia[y][x] = get_color(m)
    return pict_julia

height = 1000
width = 1000
x_1 = -2
y_1 = -2
x_2 = 2
y_2 = 2
maxi_it = 100
c = complex(-0.8,0.156)

#Affichage en blanc noir de l'ensemble de l'ensemble de Julia 
tab = julia_set(height,width,x_1,x_2,y_1,y_2)
write_picture(tab,"#Julia","Julia_en_couleur.ppm")

Image enregistrée sous Julia_en_couleur.ppm avec succès


$\textbf{IV)}$ $\underline{\textbf{ZOOM}}$

$\star\underline{\textbf{Mandelbrot}}$

Reprenons notre fonction mandelbrot_set() et nous allons zoomer sur le point de coordonnées $z=a+ib$ d'un facteur de $f=2^{30}$

In [56]:
def zoom(img:list[list[list[int]]],a:float,b:float,x_start:float,y_start:float,x_end:float,y_end:float,f:float) :
    z_start = [a - (x_end-x_start)/(2*f), b+(y_end-y_start)/(2*f)]
    z_end = [a + (x_end-x_start)/(2*f), b+(y_end-y_start)/(2*f)]
    zoom_pict = mandelbrot_set(len(img),len(img[0]),z_start[0],z_end[1],z_start[0],z_end[1])
    return zoom_pict
a_zoom = -0.7436438870371594998079545865
b_zoom = 0.13182590420531249972315198
fact = 2**30
zoom_tab = zoom(tab,a_zoom,b_zoom,x_1,y_1,x_2,y_2,fact)
write_picture(zoom_tab,"#Zoom","Mandel_zoom")


Image enregistrée sous Mandel_zoom avec succès
Le programme a mis 111.81668734550476 secondes pour s'exécuter


$\star\underline{\textbf{Julia}}$

Faisons de même avec l'ensemble de Julia sans oublier que la fonction $\textbf{julia_set()}$ ne prend en paramètre que 7 variables.

In [47]:
def zoom(img:list[list[list[int]]],a:float,b:float,x_start:float,y_start:float,x_end:float,y_end:float,f:float) :
    z_start = [a - (x_end-x_start)/(2*f), b+(y_end-y_start)/(2*f)]
    z_end = [a + (x_end-x_start)/(2*f), b+(y_end-y_start)/(2*f)]
    zoom_pict = julia_set(len(img),len(img[0]),z_start[0],z_end[1],z_start[0],z_end[1])
    return zoom_pict
a_zoom = 1
b_zoom = -0.5
fact = 2**30
zoom_tab = zoom(tab,a_zoom,b_zoom,x_1,y_1,x_2,y_2,fact)
write_picture(zoom_tab,"#Zoom","Julia_zoom")

Image enregistrée sous Julia_zoom avec succès


$\underline{\textbf{Remarque}}$

Il faut noter que le choix du max_it joue un rôle important dans la construction de la fractale

Plus le nombre d'itérations est faible plus le couleurs se distinguent très facilement cependant en augmentant le nombre d'itérations,on distingue moins bien les couleurs et c'est la qu'entre en jeu la fonction $\textbf{zoom()}$ qui permet grâce au facteur de zoom de pouvoir mieux observer et voir la constitution de la fractale.
En deuxième position il faut noter que plus la valeur de max_it est de plus en plus grande, la complexité en temps est de plus en plus grande.  

Afin de mieux l'illustrer,essayons de calculer le temps d'exécution de l'ensemble de $\textbf{mandelbrot()}$ grâce au module $\textbf{time}$

In [69]:
import time
debut = time.time()
def mandelbrot_set(height:int,width:int,x_start:float,x_end:float,y_start:float,y_end:float) :
#Création de l'image avec la couleur noir
    pict_mandelbrot = [[[0,0,0] for i in range(width)] for j in range(height)]
    
    for y in range(height):
        for x in range(width) :
            real = (x * (x_end-x_start)/(width-1))+x_start
            imag = (y * (y_end-y_start)/(height-1))+y_start
            c = complex(real,imag)
            n = mandelbrot_sub(c,maxi_it)
            #Définition de la variable k
            k = n/maxi_it
            #Affichage de l'ensemble 
            if n < maxi_it :
                pict_mandelbrot[y][x] = get_color(k)
    return pict_mandelbrot
#Entrée des valeurs     
h = 1000
w = 1000
x_1 = -2
y_1 = -2
x_2 = 2
y_2 = 2
maxi_it = 300

#création de l'image et de l'affichage en blanc noir grâce à la fonction write_picture()
tab = mandelbrot_set(h,w,x_1,x_2,y_1,y_2)
write_picture(tab,"#Mandel","Mandelbrot_en_couleur.ppm")
fin = time.time()
ecart = abs(debut - fin)
print("Le programme a mis",ecart,"secondes pour s'exécuter")

Image enregistrée sous Mandelbrot_en_couleur.ppm avec succès
Le programme a mis 19.015083074569702 secondes pour s'exécuter


$\underline{\textbf{Quelques exemples}} :$

$max_{it} = 50 \Rightarrow t_{50} = 10.60235047340393 s \\$
$max_{it} = 100 \Rightarrow t_{100} = 12.398715496063232 s \\\\ $
$max_{it} = 300 \Rightarrow t_{300} = 19.015083074569702 s$


Alors on remarque que le plus le nombre d'itérations est important plus le temps d'exécution l'est . Et il ne faut pas oublier que le nomre d'itérations joue aussi un rôle sur l'espace mémoire